src/garage/envs/bullet/bullet_env.py

```diff
@@ -25,101 +42,178 @@ def __init__(self, env=None, env_name='', is_image=False):
-                debugging, and sometimes learning)
+            TimeStep: The time step resulting from the action.
+
+        Raises:
+            RuntimeError: if `step()` is called after the environment has been
+                constructed and `reset()` has not been called.
 
         """
+        if self._last_observation is None:
+            raise RuntimeError('reset() must be called before step()!')
+
         observation, reward, done, info = self.env.step(action)
+
+        if self._visualize:
+            self.env.render(mode='human')
+
+        last_obs = self._last_observation
+        # Type conversion
+        if not isinstance(reward, float):
+            reward = float(reward)
+
+        self._last_observation = observation
+        self._step_cnt += 1
+
+        step_type = None
+        if done:
+            step_type = StepType.TERMINAL
+        elif self._step_cnt == 1:
+            step_type = StepType.FIRST
+        else:
+            step_type = StepType.MID
+
         # gym envs that are wrapped in TimeLimit wrapper modify
         # the done/termination signal to be true whenever a time
         # limit expiration occurs. The following statement sets
         # the done signal to be True only if caused by an
         # environment termination, and not a time limit
         # termination. The time limit termination signal
         # will be saved inside env_infos as
-        # 'BulletEnv.TimeLimitTerminated'
-        if 'TimeLimit.truncated' in info:
-            info['BulletEnv.TimeLimitTerminated'] = done  # done = True always
-            done = not info['TimeLimit.truncated']
-        return observation, reward, done, info
+        # 'GarageEnv.TimeLimitTerminated'
+        if 'TimeLimit.truncated' in info or \
+            self._step_cnt >= self._spec.max_episode_length:
+            info['BulletEnv.TimeLimitTerminated'] = True
+            step_type = StepType.TIMEOUT
+        else:
+            info['TimeLimit.truncated'] = False
+            info['BulletEnv.TimeLimitTerminated'] = False
+
+        return TimeStep(
+            env_spec=self.spec,
+            observation=last_obs,
+            action=action,
+            reward=reward,
+            next_observation=observation,
+            env_info=info,
+            agent_info={},  # TODO: can't be populated by env
+            step_type=step_type)
+
+    def render(self, mode):
+        """Renders the environment.
+
+        Args:
+            mode (str): the mode to render with. The string must be present in
+                `self.render_modes`.
+        """
+        if mode not in self.render_modes:
+            raise ValueError('Supported render modes are {}, but '
+                             'got render mode {} instead.'.format(
+                                 self.render_modes, mode))
+        return self.env.render(mode)
+
+    def visualize(self):
+        """Creates a visualization of the environment."""
+        self.env.render(mode='human')
```

Comment:
Does this satisfy the API contract of `visualize()`? i.e. I will call this once and then see all steps of the environment rendered

Reply:
Yes. See https://github.com/rlworkgroup/garage/blob/082c56b6557de8933012af6e31ee3e77421a3547/src/garage/envs/bullet/bullet_env.py#L142.

Gym env's `render(mode='human)` renders creates a visualization if called the first time, or else update the visualization. Verified working visualization with examples/step_env, etc.